evm/src/executor/abi/fmt.rs

```diff
@@ -0,0 +1,381 @@
+}
+
+impl FormatValue for String {
+    fn fmt(&self, spec: FormatSpec) -> String {
+        match spec {
+            FormatSpec::String => self.clone(),
+            FormatSpec::Object => format!("'{}'", self.clone()),
+            FormatSpec::Number | FormatSpec::Integer => String::from("NaN"),
+        }
+    }
+}
+
+impl FormatValue for bool {
+    fn fmt(&self, spec: FormatSpec) -> String {
+        match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number => (*self as i32).to_string(),
+            FormatSpec::Integer => String::from("NaN"),
+        }
+    }
+}
+
+impl FormatValue for U256 {
+    fn fmt(&self, _spec: FormatSpec) -> String {
+        self.pretty()
+    }
+}
+
+impl FormatValue for I256 {
+    fn fmt(&self, _spec: FormatSpec) -> String {
+        self.pretty()
+    }
+}
+
+impl FormatValue for Address {
+    fn fmt(&self, spec: FormatSpec) -> String {
+        match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number | FormatSpec::Integer => String::from("NaN"),
+        }
+    }
+}
+
+impl FormatValue for Bytes {
+    fn fmt(&self, spec: FormatSpec) -> String {
+        match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number | FormatSpec::Integer => String::from("NaN"),
+        }
+    }
+}
+
+/// Formats a `specstr` using the input values, v1, v2 and v3.
+/// For example:
+///   console_log_format_n("%s has %d characters", 2, "foo", 3) == "foo has 3 characters"
+/// num_v determines the maximum number of values to be used for the formatting (since we don't have
+/// variadics in Rust). For example: console_log_format_n("%s", 1, v1, v2, v3) will format the
+/// string using v1 only, ignoring v2 and v2.
+///
+/// Formatting rules are the same as hardhat. The supported format specifiers are as follows:
+/// - %s: Converts the value using its String representation. This is equivalent to applying
+///   UIfmt::pretty() on the format string.
+/// - %d, %i: Converts the value to an integer. If a non-numeric value, such as String or Address,
+///   is passed, then the spec is formatted as `NaN`.
+/// - %o: Treats the format value as a javascript "object" and converts it to its string
+///   representation.
+/// - %%: This is parsed as a single percent sign ('%') without consuming any input value.
+///
+/// Unformatted values are appended to the end of the formatted output using UIfmt::pretty().
+/// If there are more format specifiers than values, then the remaining unparsed format specifiers
+/// appended to the formatted output as-is.
+fn console_log_format_n(
+    specstr: &str,
+    num_v: isize,
+    v0: &dyn FormatValue,
+    v1: &dyn FormatValue,
+    v2: &dyn FormatValue,
```

Comment:
not sure if this got lost but I thought I made a comment here that this could be replaced by a single `values: impl IntoIterator<Item=&dyn FormatValue>`

so that you can get the next value like

```rust
let mut values_iter = values.into_iter();
let mut current_value = None;

let mut next_value = || {
     if let next @ Some(_) = values.next() {
        current = next;
     }
      current.as_ref().expect("non empty iterator")
};

```

Reply:
Yeah didn't see this excellent suggestion. Will do, thanks!